lib/stripe/subscriptions/subscription.ex

```diff
@@ -144,11 +144,15 @@ defmodule Stripe.Subscription do
 
   Takes the `id` and an optional map of `params`.
   """
-  @spec delete(Stripe.id() | t, Stripe.options()) :: {:ok, t} | {:error, Stripe.Error.t()}
-  def delete(id, opts \\ []) do
+  @spec delete(Stripe.id() | t, params, Stripe.options()) :: {:ok, t} | {:error, Stripe.Error.t()}
+        when params: %{
+               optional(:at_period_end) => boolean
+             }
+  def delete(id, params \\ %{}, opts \\ []) do
```

Comment:
@snewcomer Is this a breaking change. If I'm interpreting it correctly, any app that's using this endpoint as

```Elixir
id |> delete(custom_options)
``` 

will stop working with this change.

Should we add clauses to keep backwards compatibility. For example

```Elixir
def delete(id, opts) when is_list(opts), do: delete(id, %{}, opts)
def delete(id, params) when is_map(params), do: delete(id, params, [])
```

Reply:
@begedin Looking back, it seems [params](https://github.com/code-corps/stripity_stripe/commit/c95bddeebd5443fc01d6a99a50ddb3cc49cd5a27#diff-6fbf09e5bedb86d334a7690e4896ad58R219) was taken out ~8 months ago and went into 2.0-beta.  I missed it in the refactor. 

So given the time frame, it's probably a good idea to do what you recommended and perhaps add 2 tests for Subscription.delete - one passing a map and one passing a list as the second arg. 👍

Reply:
@begedin You're right! I'd forgotten about backwards compatibility.

What about:
```elixir
def delete(id, opts \\ []) when is_list(opts), do: delete(id, %{}, opts)

def delete(id, params, opts) when is_map(params) do
  # The actual thing
end
```

This way you can choose to pass `params` _and_ `opts` in a call and it's still backwards compatible. The only downside I see is that you can't pass only `params` and get the default `opts`.
So, if you want to call the function with `params` and the default `opts` it'd have to be: 

```elixir
delete(id, params, [])
```

What do you think? @begedin @snewcomer

Reply:
@strattadb Will this work?  Note - without the `when params: {` spec stuff.

```
@spec delete(Stripe.id() | t, Stripe.options()) :: {:ok, t} | {:error, Stripe.Error.t()}
def delete(id, opts \\ []) when is_list(opts), do: delete(id, %{}, opts)

@spec delete(Stripe.id() | t, map, Stripe.options()) :: {:ok, t} | {:error, Stripe.Error.t()}
def delete(id, params \\ %{}, opts \\ []) do
```

Reply:
@snewcomer  I think that's the best solution, but I tried it in IEx and it didn't compile.

```
** (CompileError) iex:3: def delete/3 defaults conflicts with delete/2
    iex:3: (module)
```

Reply:
@strattadb Oh right b/c of expansion at compile time.  I think Nikola was right in the first place 😆  (with guards)...

```
  @spec delete(Stripe.id() | t, Stripe.options()) :: {:ok, t} | {:error, Stripe.Error.t()}
  def delete(id, opts) when is_list(opts), do: delete(id, %{}, opts)

  @spec delete(Stripe.id() | t, map) :: {:ok, t} | {:error, Stripe.Error.t()}
  def delete(id, params) when is_map(params), do: delete(id, params, [])

  @spec delete(Stripe.id() | t, map, Stripe.options()) :: {:ok, t} | {:error, Stripe.Error.t()}
  def delete(id, params \\ %{}, opts \\ []) do
```

We basically need an arity of 1, 2, or 3, with the second argument being either of type List or Map.

Reply:
@snewcomer Oh, of course! I misunderstood Nikola's suggestion the first time. I'll add the clauses and tests.

Can I ask you why not the `params` spec clause?

Reply:
params spec is needed for the docs.  my bad 👌